pkg/agent/controller/bgp/controller.go

```diff
@@ -0,0 +1,1017 @@
+		},
+		resyncPeriod,
+	)
+	c.serviceInformer.AddEventHandlerWithResyncPeriod(
+		cache.ResourceEventHandlerFuncs{
+			AddFunc:    c.addService,
+			UpdateFunc: c.updateService,
+			DeleteFunc: c.deleteService,
+		},
+		resyncPeriod,
+	)
+	if c.egressEnabled {
+		c.egressInformer.AddEventHandlerWithResyncPeriod(
+			cache.ResourceEventHandlerFuncs{
+				AddFunc:    c.addEgress,
+				UpdateFunc: c.updateEgress,
+				DeleteFunc: c.deleteEgress,
+			},
+			resyncPeriod,
+		)
+	}
+	c.nodeInformer.AddEventHandlerWithResyncPeriod(
+		cache.ResourceEventHandlerFuncs{
+			AddFunc:    nil,
+			UpdateFunc: c.updateNode,
+			DeleteFunc: nil,
+		},
+		resyncPeriod,
+	)
+	if c.enabledIPv4 {
+		c.ipProtocols = append(c.ipProtocols, protocolIPv4)
+	}
+	if c.enabledIPv6 {
+		c.ipProtocols = append(c.ipProtocols, protocolIPv6)
+	}
+	return c, nil
+}
+
+// watchSecretChanges uses watch API directly to watch for the changes of the specific Secret.
+func (c *Controller) watchSecretChanges(endCh <-chan struct{}) error {
+	ns := env.GetAntreaNamespace()
+	watcher, err := c.k8sClient.CoreV1().Secrets(ns).Watch(context.TODO(), metav1.SingleObject(metav1.ObjectMeta{
+		Namespace: ns,
+		Name:      c.bgpPeerPasswordsSecret,
+	}))
+	if err != nil {
+		return fmt.Errorf("failed to create Secret watcher: %v", err)
+	}
+
+	ch := watcher.ResultChan()
+	defer watcher.Stop()
+	klog.InfoS("Starting watching Secret changes", "Secret", fmt.Sprintf("%s/%s", ns, c.bgpPeerPasswordsSecret))
+	for {
+		select {
+		case event, ok := <-ch:
+			if !ok {
+				return nil
+			}
+			// Update BGP peer passwords.
+			klog.InfoS("Processing Secret event", "Secret", fmt.Sprintf("%s/%s", ns, c.bgpPeerPasswordsSecret))
+			func() {
+				c.bgpPeerPasswordsMutex.Lock()
+				defer c.bgpPeerPasswordsMutex.Unlock()
+
+				secretObj := event.Object.(*corev1.Secret)
+				c.bgpPeerPasswords = make(map[string]string)
+				for key, data := range secretObj.Data {
+					c.bgpPeerPasswords[key] = string(data)
+				}
+			}()
+			func() {
+				c.bgpPolicyBindingMutex.RLock()
+				defer c.bgpPolicyBindingMutex.RUnlock()
+				if c.bgpPolicyBinding.effectiveBP != "" {
+					c.queue.Add(c.bgpPolicyBinding.effectiveBP)
+				}
+			}()
+		case <-endCh:
+			return nil
+		}
+	}
+}
+
+func (c *Controller) Run(stopCh <-chan struct{}) {
+	defer c.queue.ShutDown()
+
+	klog.InfoS("Starting", "controllerName", controllerName)
+	defer klog.InfoS("Shutting down", "controllerName", controllerName)
+
+	if !cache.WaitForNamedCacheSync(controllerName,
+		stopCh,
+		c.nodeListerSynced,
+		c.serviceListerSynced,
+		c.egressListerSynced,
+		c.bgpPolicyListerSynced,
+		c.endpointSliceListerSynced) {
+		return
+	}
+
+	go wait.NonSlidingUntil(func() {
+		if err := c.watchSecretChanges(stopCh); err != nil {
+			klog.ErrorS(err, "Watch Secret error", "secret", c.bgpPeerPasswordsSecret)
+		}
+	}, time.Second*10, stopCh)
+
+	for i := 0; i < defaultWorkers; i++ {
+		go wait.Until(c.worker, time.Second, stopCh)
+	}
```

Comment:
There is no point to process multiple policies in parallel and have many code to handle the race condition.
In normal scenario, we expect only zero or single policy applied to the Node, we should just get one policy and implement it, instead of dealing with multiple policies like it's a normal case. If all jobs of implementing a single policy should be done sequentially, it just needs a single worker with a dummy key.

Reply:
The worker's implementation in my mind:
```go
func syncBGPPolicy() error {
  // Get 0 or 1 policy applied to the Node.
  bgpPolicy, exists := c.getBGPPolicy() 
  if !exists {
    // cleanup
  }
  // Implement the bgpPolicy
}
```

Reply:
How should we choose a BGPPolicy if users apply multiple ones to a Node anyway?